### PR TITLE
Add Translatable do BackOffice

### PR DESCRIPTION
**Pull request by pedro-c-oliveira**
Add Translatable do BackOffice.
Only work on create table. Don't work on edit table. 
Made editions manually on model

### CODE REVIEW

**Comment by SteamWind** [comment]
You shouldn't add the translation of the language you can't translate...

**Comment by fletch3555** [comment]
On the contrary @SteamWind, I'd prefer new translation keys be added to all languages with English as a default.  This will keep the files up to date.  It looks better to have English text in a couple places than `new.translation.string`.

As for this PR, I'm not sure I understand what this is supposed to do...

**Comment by SteamWind** [comment]
Okay, put the key is a good thing but what if we want to change the text of the english file? We need to change it everywhere then.
(I speak in general case, in this one it should be okay)

**Comment by fletch3555** [comment]
Only the languages you use. And you'd likely have to do that even if it was translated appropriately.

File: publishable/lang/fr/database.php [hunk 1]
```diff
@@ -62,6 +62,7 @@
     'url_slug'                  => 'URL Slug (doit être unique)',
     'url_slug_ph'               => 'URL slug (exemple : articles)',
     'visibility'                => 'Visibilité',
+    'is_translatable'           => 'Do you wish install translations?',
```
Comment (SteamWind) [inline publishable/lang/fr/database.php]: `'is_translatable'           => 'Voulez-vous installer les traductions?',`

File: publishable/lang/fr/generic.php [hunk 2]
```diff
@@ -87,4 +87,5 @@
     'viewing'                => 'Affichage',
     'yes'                    => 'Oui',
     'yes_please'             => 'Oui SVP',
+    'is_translatable'        => 'Translatate?',
```
Comment (SteamWind) [inline publishable/lang/fr/generic.php]: `'is_translatable'        => 'Traduire?',`

**Review by SteamWind — CHANGES_REQUESTED** [review]

**Comment by pedro-c-oliveira** [comment]
This is what this pull enhancement do in backoffice:

1. When do you wish that voyager creates model for your table, shows a button ask if you wish that model could be translatable.


2. if you wish apply translatable to your model a new column apear on table fields to choose wich fields are translatable.




3. After that you can check your new model and you can see the trait and fields translatable array on php file.




At the moment this feature only works on create new table not in edit mode.

File: publishable/lang/fr/generic.php (hunk 2 above)
Comment (pedro-c-oliveira) [reply publishable/lang/fr/generic.php]: i just add the necessary keys for all language files in voyager on english, after that, people will translate the key for the correct sentence in his language.

**Comment by Martins-74** [comment]
This is very useful, we can make multilanguages websites in minutes with this enhancement.
Hope it get's approved  👍 ❤️

File: src/Commands/MakeModelCommand.php [hunk 3]
```diff
@@ -67,6 +67,35 @@ protected function addSoftDelete(&$stub)
...
+        $traitIncl = $trait = $translatablefields = '';
+
+        if ($this->option('traitTranslatable')) {
+            $traitIncl = 'use TCG\Voyager\Traits\Translatable;';
+            $trait = 'use Translatable;';
+            $translatablefields='protected $translatable = [';
+            foreach ($this->option('traitTranslatable') as $index => $field) {
+                $translatablefields.='"'.$field.'",';
+            }
+            $translatablefields.='];';
```
Comment (emptynick) [inline src/Commands/MakeModelCommand.php]:
```
$translatablefields = 'protected $translatable = ["'.implode($this->option('traitTranslatable'), '", "').'"]';
```
Saves you 4 lines.

**Review by emptynick — CHANGES_REQUESTED** [review]

File: src/Commands/MakeModelCommand.php (hunk 3 above)
Comment (pedro-c-oliveira) [reply src/Commands/MakeModelCommand.php]: thank you

File: publishable/lang/fr/generic.php (hunk 2 above)
Comment (SteamWind) [reply publishable/lang/fr/generic.php]: Yes okay.

**Comment by SteamWind** [comment]
Hope we could have this in edit mode soon :)

**Review by pedro-c-oliveira — COMMENTED** [review]
improve code efficience

File: src/Commands/MakeModelCommand.php (hunk 3 above)
Comment (pedro-c-oliveira) [reply src/Commands/MakeModelCommand.php]: request changes done.

**Comment by pedro-c-oliveira** [comment]
Someone reviewer to aprove the pull please

File: src/Commands/MakeModelCommand.php [hunk 4]
```diff
@@ -67,6 +67,30 @@ protected function addSoftDelete(&$stub)
...
     }
 
+    /**
+     * Add Translatable to the given stub.
+     *
+     * @param string $stub
+     *
+     * @return $this
+     */
+    protected function addTraitTranslatable(&$stub)
```
Comment (fletch3555) [inline src/Commands/MakeModelCommand.php]: Can you rename the method so it's consistent with the others?

**Review by fletch3555 — CHANGES_REQUESTED** [review]

**Comment by SteamWind** [comment]
Does it then use the translation table?

File: src/Commands/MakeModelCommand.php (hunk 4 above)
Comment (pedro-c-oliveira) [reply src/Commands/MakeModelCommand.php]: @fletch3555 do you prefer just "addTranslatable" ?  or suggest another name?

**Comment by emptynick** [comment]
I just fixed the merge-conflicts.
You have some nasty JS scripts to show/hide the checkboxes, but the database editor is a vue-app which would make them unnecessary if used properly.

**Comment by emptynick** [comment]
I'm going to close this PR because it has changes requested and wasn't touched for over half a year.
If you are still interested in fixing the requested changes, please let us know.

**Pull request closed without merging**